src/utilities/expiration_map.rs

```diff
@@ -0,0 +1,210 @@
+// Licenses can be found in the root directory of this project at LICENSE, COPYING and CONTRIBUTOR.
+//
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+/// A time limited map of key-value pairs.
+#[allow(unused)]
+pub struct ExpirationMap<K, V> {
+    map: ::std::collections::BTreeMap<K, (V, ::time::SteadyTime)>,
+    time_to_live: ::time::Duration,
+}
+
+#[allow(unused)] // remove when an ExpirationMap has been included elsewhere in the code
+impl<K, V> ExpirationMap<K, V> where K: PartialOrd + Ord + Clone, V: Clone {
+    /// Constructor
+    pub fn with_expiry_duration(time_to_live: ::time::Duration) -> ExpirationMap<K, V> {
+        ExpirationMap {
+            map: ::std::collections::BTreeMap::new(),
+            time_to_live: time_to_live,
+        }
+    }
+
+    /// Inserts a key-value pair into the map. Returns replaced value if key is already present.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        match self.map.insert(key, (value, ::time::SteadyTime::now())) {
+            Some((value, _)) => Some(value),
+            None => None,
+        }
+    }
+
+    /// If key exists remove it from the map and return corresponding value, whether expired or not.
+    pub fn remove(&mut self, key: &K) -> Option<V> {
+        match self.map.remove(key) {
+            Some((value, _)) => Some(value),
+            None => None,
+        }
+    }
+
+    /// Retrieves a value for the given key if present in the map.
+    pub fn get(&mut self, key: &K) -> Option<&V> {
+        match self.map.get(key) {
+            Some(&(ref value, time)) => {
+                if time + self.time_to_live < ::time::SteadyTime::now() {
+                    None
+                } else {
+                    Some(value)
+                }
+            },
+            None => None
+        }
+    }
+
+    /// Returns true if a value exists for the specified key.
+    pub fn contains_key(&mut self, key: &K) -> bool {
+        match self.map.get(key) {
+            Some(&(_, time)) => {
+                if time + self.time_to_live < ::time::SteadyTime::now() {
+                    false
+                } else {
+                    true
+                }
+            },
+            None => false
+        }
+    }
+
+    /// Recover expired key-value pairs removing any such from the map.
+    pub fn remove_expired(&mut self) -> Option<Vec<(K,V)>> {
+        let mut expired = Vec::new();
+
+        for (key, &(ref value, time)) in self.map.iter() {
+            if time + self.time_to_live < ::time::SteadyTime::now() {
```

Comment:
probably better to only call ::time::SteadyTime::now() only once, but good for now